frontend/lib/resolution/try-catch-analysis.cpp

```diff
@@ -0,0 +1,376 @@
+    const AstNode* ast;
+    int  tryDepth;
+    bool fnDoesThrow;
+    int taskFunctionDepth;
+  };
+
+  struct TryCatchAnalyzer {
+    using RV = ResolvedVisitor<TryCatchAnalyzer>;
+
+    // input
+    Context* context;
+    const AstNode* astForErr;
+    errorCheckingMode mode;
+    // need to keep a stack of current ast and some meta info
+    // like if we're in a place it's ok to throw, what the current
+    // error checking mode is, etc.
+    std::vector<TryCatchState> tryStack;
+    int tryDepth = 0;
+    // output
+    // TODO: figure out the output - errors when something doesn't pass?
+    // maybe no output?
+    std::vector<QualifiedType> returnedTypes;
+
+    TryCatchAnalyzer(Context* context,
+                  const AstNode* symbol,
+                  errorCheckingMode mode)
+        : context(context),
+          astForErr(symbol),
+          mode(mode) {
+    }
+
+    void checkCatches(const Try* node, RV& rv) {
+
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (hasCatchAll) {
+          context->error(node, "catchall placed before the end of a catch list");
+        }
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+        }
+      }
+    }
+
+    // Returns true if the catch is a catchall statement.
+    bool isCatchAll(const Catch* node, RV& rv) {
+        bool catchAll = false;
+        if (auto err = node->error()) {
+          // we validate that the type of error is a subtype of Error or Error
+          // itself resolving the AST.
+          // So here we just need to check for the base Errors.Error to know if
+          // this is a catchall.
+          const ResolvedExpression& r = rv.byAst(err);
+          auto errType = CompositeType::getErrorType(context);
+          auto dec = ClassTypeDecorator(ClassTypeDecorator::MANAGED_NONNIL);
+          auto manager = AnyOwnedType::get(context);
+          auto ret = ClassType::get(context, errType->basicClassType(),
+                                    manager, dec);
+          auto qt = QualifiedType(QualifiedType::VAR,
+                                  ret->withDecorator(context, dec));
+          if (r.type() == qt) {
+            catchAll = true;
+          }
+        } else {
+          // this is a catchall since it has no error associated with it
+          catchAll = true;
+        }
+        return catchAll;
+    }
+
+    // Returns true if the catches don't cover all of the cases.
+     bool catchesNotExhaustive(const Try* node, RV& rv) {
+      if (node->isTryBang()) {
+        return false;
+      }
+
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+          break;
+        }
+      }
+
+      return !hasCatchAll;
+    }
+
+    void enterAst(const AstNode* ast, bool doesThrow) {
+        TryCatchState tca;
+        tca.ast = ast;
+        tca.tryDepth = tryStack.size();
```

Comment:
This is never used, and, if you're only ever accessing the top of the try stack (as you seem to be), you can always just use the tryStack directly.